File: devel/darts/darray.py
'''
File: darray.py
This is Darray smart string searcher
File Created: Wednesday, 24th June 2020 6:05:43 pm
Author: enxu (xuen@mokar.com)
-----
Last Modified: Wednesday, 24th June 2020 6:07:47 pm
Modified By: enxu (xuen@mokahr.com)
-----
Copyright 2021 - 2020 Your Company, Moka
'''
__all__ = ['compile']


class _Trie():
    def __init__(self, check, base, fail, output, v, l, codeMap):
        self.check = check
        self.base = base
        self.fail = fail
        self.output = output
        self.v = v
        self.l = l
        self.codemap = codeMap
        self.maxlen = max(self.l)+1

    def _getCode(self, seq):
        seq = str(seq).lower()
        return self.codemap.get(seq, len(self.codemap)+1)

    def parse(self, text):
        currentState = 0
        for position, seq in enumerate(text):
            position += 1
            currentState = self._getState(currentState, self._getCode(seq))
            hitArray = self.output[currentState]
            if hitArray is not None:
                for hit in hitArray:
                    yield(position - self.l[hit], position, self.v[hit])

    def parseWithSkip(self, text, skipCall):
        currentState = 0
        index_bufer = [-1]*self.maxlen
        index_buffer_pos = 0
        for position, seq in enumerate(text):
            if skipCall(seq):
                continue
            index_bufer[index_buffer_pos % self.maxlen] = position
            index_buffer_pos += 1
            currentState = self._getState(currentState, self._getCode(seq))
            hitArray = self.output[currentState]
            if hitArray is not None:
                for hit in hitArray:
                    pre_index = (index_buffer_pos-self.l[hit]) % self.maxlen
                    yield(index_bufer[pre_index], position+1, self.v[hit])

    def _getState(self, currentState, character):
        newCurrentState = self._transitionWithRoot(currentState, character)
        while newCurrentState == -1:
            currentState = self.fail[currentState]
            newCurrentState = self._transitionWithRoot(currentState, character)
        return newCurrentState

    def _transitionWithRoot(self, nodePos,  c):
        b = self.base[nodePos]
        p = b + c + 1
        if b != self.check[p]:
            if nodePos == 0:
                return 0
            return -1
        return p


class _State():
    def __init__(self, depth=0):
        self.depth = depth
        self.failure = None
        self.emits = set()
        self.success = {}
        self.index = 0

    def getLargestValueId(self):
        if not self.emits:
            return None
        return max(self.emits)

    def isAcceptable(self):
        return self.depth > 0 and len(self.emits) > 0

    def setFailure(self, failState, fail):
        self.failure = failState
        fail[self.index] = failState.index

    def nextState(self, character, ignoreRootState=False):
        nextState = self.success.get(character, None)
        if (not ignoreRootState) and (nextState is None) and (self.depth == 0):
            nextState = self
        return nextState

    def addState(self, character):
        nextS = self.nextState(character, True)
        if nextS is None:
            nextS = _State(self.depth + 1)
            self.success[character] = nextS
        return nextS

    def getStates(self):
        return iter(v for (k, v) in sorted(self.success.items(), key=lambda x: x[0]))

    def getTransitions(self):
        return sorted(self.success.keys())


class _Builder():
    def __init__(self):
        self.rootState = _State()
        self.used = []
        self.base = []
        self.check = []
        self.v = None
        self.codemap = {}
        self.l = None
        self.allocSize = 0
        self.progress = 0
        self.nextCheckPos = 0
        self.size = 0

    def build(self, kvs, strKey):
        max_code = self._addAllKeyword(kvs, strKey)
        # build double array tire base on tire
        self._buildDoubleArrayTrie(max_code)
        self.used = None
        # build failure table and merge output table
        self._constructFailureStates()
        self.rootState = None
        self._loseWeight()
        return _Trie(self.check, self.base, self.fail, self.output, self.v, self.l, self.codemap)

    def resize(self, newSize):
        base2 = [0]*newSize
        check2 = [0]*newSize
        used2 = [False]*newSize
        if self.allocSize > 0:
            base2[:self.allocSize] = self.base[:self.allocSize]
            check2[:self.allocSize] = self.check[:self.allocSize]
            used2[:self.allocSize] = self.used[:self.allocSize]

        self.base = base2
        self.check = check2
        self.used = used2
        self.allocSize = newSize
        return self.allocSize

    def _addAllKeyword(self, kvs, strKey):
        max_code = 0
        self.l = []
        self.v = []
        for index, kv in enumerate(kvs):
            currentState = self.rootState
            lens = 0
            for seq in kv[0]:
                seq = strKey(seq)
                if seq is None:
                    continue
                lens += 1
                _seq = str(seq).lower()
                _code = self.codemap.get(_seq, len(self.codemap)+1)
                self.codemap[_seq] = _code
                currentState = currentState.addState(_code)
            if lens < 1:
                continue
            currentState.emits.add(index)
            self.l.append(lens)
            max_code += lens
            self.v.append(kv[1])
        return int((max_code+len(self.codemap))/1.5) + 1

    def _fetch(self, parent):
        siblings = []
        if parent.isAcceptable():
            fakeNode = _State(-(parent.depth + 1))
            fakeNode.emits.add(parent.getLargestValueId())
            siblings.append((0, fakeNode))
        for (k, v) in sorted(parent.success.items(), key=lambda x: x[0]):
            siblings.append((k + 1, v))
        return siblings

    def _insert(self, siblings):
        begin = 0
        pos = max(siblings[0][0] + 1, self.nextCheckPos) - 1
        nonzero_num = 0
        first = 0
        if self.allocSize <= pos:
            self.resize(pos + 1)
        while True:
            pos += 1
            if self.allocSize <= pos:
                self.resize(pos + 1)
            if self.check[pos] != 0:
                nonzero_num += 1
                continue
            elif first == 0:
                nextCheckPos = pos
                first = 1

            begin = pos - siblings[0][0]
            if self.allocSize <= (begin + siblings[-1][0]):
                self.resize(begin + siblings[-1][0]+100)
            if self.used[begin]:
                continue
            flag = False
            for kv in siblings:
                if self.check[begin + kv[0]] != 0:
                    flag = True
                    break

            if not flag:
                break

        if 1.0 * nonzero_num/(pos - nextCheckPos + 1) >= 0.95:
            nextCheckPos = pos
        self.used[begin] = True

        if self.size < begin + siblings[-1][0] + 1:
            self.size = begin + siblings[-1][0] + 1

        for kv in siblings:
            self.check[begin + kv[0]] = begin

        for kv in siblings:
            new_siblings = self._fetch(kv[1])
            if not new_siblings:
                self.base[begin + kv[0]] = -(kv[1].getLargestValueId()+1)
                self.progress += 1
            else:
                self.base[begin + kv[0]] = self._insert(new_siblings)
            kv[1].index = begin + kv[0]
        return begin

    def _buildDoubleArrayTrie(self, max_code):
        self.progress = 0
        self.resize(max_code+10)
        self.base[0] = 1
        self.nextCheckPos = 0
        root_node = self.rootState
        siblings = self._fetch(root_node)
        if not siblings:
            return
        self._insert(siblings)

    def _constructFailureStates(self):
        self.fail = [0]*(self.size+1)
        self.output = [None]*(self.size+1)
        queue = []

        for depthOneState in self.rootState.getStates():
            depthOneState.setFailure(self.rootState, self.fail)
            queue.append(depthOneState)
            self._constructOutput(depthOneState)

        while len(queue) > 0:
            currentState = queue.pop(0)
            for transition in currentState.getTransitions():
                targetState = currentState.nextState(transition)
                queue.append(targetState)
                traceFailureState = currentState.failure
                while traceFailureState.nextState(transition) is None:
                    traceFailureState = traceFailureState.failure
                newFailureState = traceFailureState.nextState(transition)
                targetState.setFailure(newFailureState, self.fail)
                targetState.emits |= newFailureState.emits
                self._constructOutput(targetState)

    def _constructOutput(self, targetState):
        emit = targetState.emits
        if (emit is None) or len(emit) == 0:
            return
        self.output[targetState.index] = list(emit)

    def _loseWeight(self):
        nbase = [0]*(self.size+len(self.codemap)+1)
        nbase[:self.size] = self.base[:self.size]
        self.base = nbase

        ncheck = [0]*(self.size + len(self.codemap) + 1)
        ncheck[:self.size] = self.check[:self.size]
        self.check = ncheck


def compile(kv_iter, strKey=None):
    if strKey is None:
        def strKey(x): return x
    return _Builder().build(kv_iter, strKey)

File: src/darts/darray.go
package darts

/*
 * File: darray.go
 * Project: darts
 * File Created: Thursday, 31st December 2020 11:07:30 pm
 * Author: enxu (xuen@mokar.com)
 * -----
 * Last Modified: Thursday, 31st December 2020 11:07:37 pm
 * Modified By: enxu (xuen@mokahr.com)
 * -----
 * Copyright 2021 - 2020 Your Company, Moka
 */

import (
	"container/list"
	"encoding/gob"
	"io"
	"sort"

	"github.com/emirpasic/gods/maps/treemap"
)

//StringIter is Used for parer input
type StringIter func(func(str *string, postion int) bool)

//StrLabelIter is a input
type StrLabelIter func(func(strlist StringIter, labels []int) bool)

//Trie is a double array trie
type Trie struct {
	Check, Base, Fail, L []int
	V                    [][]int
	OutPut               [][]int
	MaxLen               int
	CodeMap              map[string]int
}

//WriteToBytes is write the trie to bytes
func (t *Trie) WriteToBytes(writer io.Writer) error {
	return gob.NewEncoder(writer).Encode(t)
}

//ReadFromBytes read data
func (t *Trie) ReadFromBytes(reader io.Reader) error {
	return gob.NewDecoder(reader).Decode(t)
}

func newTrie() *Trie {
	t := new(Trie)
	t.CodeMap = make(map[string]int)
	return t
}

// get string code
func (t *Trie) getcode(word *string) int {
	code, ok := t.CodeMap[*word]
	if !ok {
		return len(t.CodeMap) + 1
	}
	return code
}

//trans
func (t *Trie) transitionWithRoot(nodePos int, c int) int {
	b := 0
	if nodePos < len(t.Base) {
		b = t.Base[nodePos]
	}
	p := b + c + 1
	x := 0
	if p < len(t.Check) {
		x = t.Check[p]
	}
	if b != x {
		if nodePos == 0 {
			return 0
		}
		return -1
	}
	return p
}

// get state transpose
func (t *Trie) getstate(currentState int, character int) int {
	newCurrentState := t.transitionWithRoot(currentState, character)
	for newCurrentState == -1 {
		currentState = t.Fail[currentState]
		newCurrentState = t.transitionWithRoot(currentState, character)
	}
	return newCurrentState
}

//ParseText parse a text list hit ( [start,end),tagidx)
func (t *Trie) ParseText(text StringIter, hit func(int, int, []int) bool) {
	currentState, indexBufferPos := 0, 0
	indexBufer := make([]int, t.MaxLen)
	text(func(seq *string, position int) bool {
		indexBufer[indexBufferPos%t.MaxLen] = position
		indexBufferPos++
		currentState = t.getstate(currentState, t.getcode(seq))
		hitArray := t.OutPut[currentState]
		for _, h := range hitArray {
			preIndex := (indexBufferPos - t.L[h]) % t.MaxLen
			val := hit(indexBufer[preIndex], position+1, t.V[h])
			if val {
				return true
			}
		}
		return false
	})

}

//State is DFA mechine state
type State struct {
	depth   int          // the string length
	failure *State       // match failed use
	emits   []int        // emits
	success *treemap.Map //go map
	index   int          //index of the struct
}

func newState(depth int) *State {
	S := new(State)
	S.depth = depth
	S.success = treemap.NewWithIntComparator()
	return S
}

//insert sort a sort to keep it order
func insertSorted(s []int, e int) []int {
	i := sort.Search(len(s), func(i int) bool { return (s)[i] <= e })
	if i == len(s) {
		s = append(s, e)
		return s
	}
	if s[i] == e {
		return s
	}
	s = append(s, 0)
	copy(s[i+1:], s[i:])
	s[i] = e
	return s
}

//add emit
func (s *State) addEmit(keyword int) {
	if keyword != -100000 {
		s.emits = insertSorted(s.emits, keyword)
	}
}

//get l code
func (s *State) getMaxValueIDgo() int {
	if len(s.emits) < 1 {
		return -100000
	}
	return s.emits[0]
}

func (s *State) isAcceptable() bool {
	return s.depth > 0 && len(s.emits) > 0
}

func (s *State) setFailure(failState *State, fail []int) {
	s.failure = failState
	fail[s.index] = failState.index
}

func (s *State) nextState(character int, ignoreRootState bool) *State {
	nextState, _ := s.success.Get(character)

	if (!ignoreRootState) && (nextState == nil) && (s.depth == 0) {
		nextState = s
	}
	if nextState == nil {
		return nil
	}
	return nextState.(*State)
}

func (s *State) addState(character int) *State {
	nextS := s.nextState(character, true)
	if nextS == nil {
		nextS = newState(s.depth + 1)
		s.success.Put(character, nextS)
	}
	return nextS
}

//Builder is inner useed
type Builder struct {
	rootState *State
	trie      *Trie
	/**
	* whether the position has been used
	 */
	used []bool
	/**
	* the allocSize of the dynamic array
	 */
	allocSize int
	/**
	* the next position to check unused memory
	 */
	nextCheckPos int
	/**
	* the size of the key-pair sets
	 */
	size int
}

//zip no use data
func (b *Builder) zipWeight() {
	msize := b.size
	newBase := make([]int, msize)
	copy(newBase[:b.size], b.trie.Base[:b.size])
	b.trie.Base = newBase

	newCheck := make([]int, msize, msize)
	copy(newCheck[:b.size], b.trie.Check[:b.size])
	b.trie.Check = newCheck
}

func (b *Builder) constructFailureStates() {
	b.trie.Fail = make([]int, b.size+1)
	b.trie.OutPut = make([][]int, b.size+1)
	queue := list.New()

	cpy := func(ori []int) []int {
		dest := make([]int, len(ori))
		copy(dest, ori)
		return dest
	}

	for _, state := range b.rootState.success.Values() {
		depthOneState := state.(*State)
		depthOneState.setFailure(b.rootState, b.trie.Fail)
		queue.PushBack(state)
		if len(depthOneState.emits) > 0 {
			b.trie.OutPut[depthOneState.index] = cpy(depthOneState.emits)
		}

	}
	for queue.Len() > 0 {
		currentState := queue.Remove(queue.Front()).(*State)
		for _, key := range currentState.success.Keys() {
			transition := key.(int)
			targetState := currentState.nextState(transition, false)
			queue.PushBack(targetState)
			traceFailureState := currentState.failure
			for traceFailureState.nextState(transition, false) == nil {
				traceFailureState = traceFailureState.failure
			}
			newFailureState := traceFailureState.nextState(transition, false)
			targetState.setFailure(newFailureState, b.trie.Fail)
			for _, e := range newFailureState.emits {
				targetState.addEmit(e)
			}
			b.trie.OutPut[targetState.index] = cpy(targetState.emits)
		}
	}
}

func (b *Builder) addAllKeyword(kvs StrLabelIter) int {
	maxCode, index, t := 0, -1, b.trie
	kvs(func(k StringIter, v []int) bool {
		index++
		lens := 0
		currentState := b.rootState
		k(func(s *string, _ int) bool {
			lens++
			code := b.trie.getcode(s)
			t.CodeMap[*s] = code
			currentState = currentState.addState(code)
			return false
		})
		currentState.addEmit(index)
		t.L = append(t.L, lens)
		if lens > t.MaxLen {
			t.MaxLen = lens
		}
		maxCode += lens
		t.V = append(t.V, v)

		return false
	})
	t.MaxLen++
	return int(float32(maxCode+len(t.CodeMap))/1.5) + 1
}

//resize data
func (b *Builder) resize(newSize int) {
	base2 := make([]int, newSize)
	check2 := make([]int, newSize)
	used2 := make([]bool, newSize)
	if b.allocSize > 0 {
		copy(base2[:int(b.allocSize)], b.trie.Base[:int(b.allocSize)])
		copy(check2[:int(b.allocSize)], b.trie.Check[:int(b.allocSize)])
		copy(used2[:int(b.allocSize)], b.used[:int(b.allocSize)])
	}
	b.trie.Base = base2
	b.trie.Check = check2
	b.used = used2
	b.allocSize = newSize
}

//Pair si tmp used
type Pair struct {
	K interface{}
	V interface{}
}

func fetch(parent *State) []Pair {
	siblings := make([]Pair, 0, parent.success.Size()+1)
	if parent.isAcceptable() {
		fakeNode := newState(-(parent.depth + 1))
		fakeNode.addEmit(parent.getMaxValueIDgo())
		siblings = append(siblings, Pair{0, fakeNode})
	}
	it := parent.success.Iterator()
	for it.Next() {
		siblings = append(siblings, Pair{it.Key().(int) + 1, it.Value()})
	}
	return siblings
}

func (b *Builder) insert(queue *list.List) {
	tCurrent := queue.Remove(queue.Front()).(Pair)
	value, siblings := tCurrent.K.(int), tCurrent.V.([]Pair)

	begin, nonZeroNum := 0, 0
	first := true
	pos := b.nextCheckPos - 1
	if pos < siblings[0].K.(int) {
		pos = siblings[0].K.(int)
	}
	if b.allocSize <= pos {
		b.resize(pos + 1)
	}
	t := b.trie
	for {
		pos++
		if b.allocSize <= pos {
			b.resize(pos + 1)
		}
		if t.Check[pos] != 0 {
			nonZeroNum++
			continue
		} else if first {
			b.nextCheckPos = pos
			first = false
		}

		begin = pos - siblings[0].K.(int)
		if b.allocSize <= (begin + siblings[len(siblings)-1].K.(int)) {
			b.resize(begin + siblings[len(siblings)-1].K.(int) + 100)
		}
		if b.used[begin] {
			continue
		}
		allIszero := true
		for i := 0; i < len(siblings); i++ {
			if t.Check[begin+siblings[i].K.(int)] != 0 {
				allIszero = false
				break
			}
		}
		if allIszero {
			break
		}
	}

	if float32(nonZeroNum)/float32(pos-b.nextCheckPos+1) >= 0.95 {
		b.nextCheckPos = pos
	}
	b.used[begin] = true

	if b.size < begin+siblings[len(siblings)-1].K.(int)+1 {
		b.size = begin + siblings[len(siblings)-1].K.(int) + 1
	}
	for i := 0; i < len(siblings); i++ {
		t.Check[begin+siblings[i].K.(int)] = begin
	}
	for i := 0; i < len(siblings); i++ {
		kv := siblings[i]
		newSiblings := fetch(kv.V.(*State))
		if len(newSiblings) < 1 {
			t.Base[begin+kv.K.(int)] = -(kv.V.(*State).getMaxValueIDgo() + 1)
		} else {
			queue.PushBack(Pair{begin + kv.K.(int), newSiblings})
		}
		kv.V.(*State).index = begin + kv.K.(int)
	}
	if value >= 0 {
		t.Base[value] = begin
	}
}

func (b *Builder) build(kvs StrLabelIter) {
	maxCode := b.addAllKeyword(kvs)
	//build double array tire base on tire
	b.resize(maxCode + 10)
	b.trie.Base[0] = 1
	siblings := fetch(b.rootState)
	if len(siblings) > 0 {
		queue := list.New()
		queue.PushBack(Pair{-1, siblings})
		for queue.Len() > 0 {
			b.insert(queue)
		}
	}
	//build failure table and merge output table
	b.constructFailureStates()
	b.rootState = nil
	b.zipWeight()
}

//CompileTrie the data
func CompileTrie(kvs StrLabelIter) *Trie {
	var builder Builder
	builder.rootState = newState(0)
	builder.trie = newTrie()
	builder.build(kvs)
	//clear mem
	result := builder.trie
	builder.trie = nil
	return result
}
